Document lang items (#978)
rust-lang/rustc-dev-guide: src/lang-items.md

# Lang items

The `rustc` compiler has certain pluggable operations, that is,
functionality that isn't hard-coded into the language, but is
implemented in libraries, with a special marker to tell the compiler
it exists. The marker is the attribute `#[lang = "..."]` and there are
various different values of `...`, i.e. various different 'lang
items'.

Many such lang items can be implemented only in one sensible way, such as
`add` (`trait core::ops::Add`) or `future_trait`
(`trait core::future::Future`). 

> Others can be overriden to achieve some
> specific goals.
> 
> For example, later sections describe how to control
> your binary startup or override panic implementation.

Reviewer: This is no longer true.
```suggestion
Others can be overridden to achieve some specific goals; for example, you can control your binary's entrypoint.
```

Features provided by lang items include:

- overloadable operators via traits: the traits corresponding to the
  `==`, `<`, dereferencing (`*`) and `+` (etc.) operators are all
  marked with lang items; those specific four are `eq`, `ord`,
  `deref`, and `add` respectively.
- stack unwinding and general failure; the `eh_personality`,
  `panic` and `panic_bounds_checks` lang items.

> - the traits in `std::marker` used to indicate types of
>   various kinds; lang items `send`, `sync` and `copy`.

Reviewer: ```suggestion
- the traits in `std::marker` used to indicate properties of
  types that are used by the compiler; lang items `send`, `sync` and `copy`.
```

  - `div_assign`: `library/core/src/ops/arith.rs`
  - `rem_assign`: `library/core/src/ops/arith.rs`
  - `eq`: `library/core/src/cmp.rs`
  - `ord`: `library/core/src/cmp.rs`
- Functions
  - `fn`: `library/core/src/ops/function.rs`
  - `fn_mut`: `library/core/src/ops/function.rs`
  - `fn_once`: `library/core/src/ops/function.rs`
  - `generator_state`: `library/core/src/ops/generator.rs`
  - `generator`: `library/core/src/ops/generator.rs`
- Opting out
  - `unsafe_cell` (relaxes pointer provenance rules, allowing const-to-mut
    casts): `library/core/src/cell.rs`
  - `manually_drop` (opts out of implicit destructor call): `library/core/src/mem/manually_drop.rs`
- Other
  - `coerce_unsized`: `library/core/src/ops/unsize.rs`
  - `drop`: `library/core/src/ops/drop.rs`
  - `drop_in_place`: `library/core/src/ptr.rs`
  - `clone`: `library/core/src/clone.rs`
  - `copy`: `library/core/src/marker.rs`
  - `send`: `library/core/src/marker.rs`
  - `sized`: `libcrary/ore/msrc/arker.rs`
  - `unsize`: `library/core/src/marker.rs`
  - `sync`: `library/core/src/marker.rs`
  - `phantom_data`: `libcore/marker.rs`
  - `discriminant_kind`: `library/core/src/marker.rs`

>   - `freeze`: `library/core/src/marker.rs`

Reviewer: I still think this is not a good fit because it will get outdated so rapidly. Maybe instead you could recommend `rg '#\[lang' library` or looking at https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/lang_items/enum.LangItem.html ?

Reviewer 2: Yeah, I was going to suggest that it would be better to have people look at the docs. And suggesting `rg -F '#[lang' library/` is probably a good idea too.

with `#![no_main]` and then create the appropriate symbol with the
correct ABI and the correct name, which requires overriding the
compiler's name mangling too:

```rust,ignore
#![feature(lang_items, core_intrinsics)]
#![feature(start)]
#![no_std]
#![no_main]
use core::intrinsics;
use core::panic::PanicInfo;

// Pull in the system libc library for what crt0.o likely requires.
extern crate libc;

// Entry point for this program.
#[no_mangle] // ensure that this symbol is called `main` in the output
pub extern fn main(_argc: i32, _argv: *const *const u8) -> i32 {
    0
}

// These functions are used by the compiler, but not
// for a bare-bones hello world. These are normally
// provided by libstd.
#[lang = "eh_personality"]
#[no_mangle]
pub extern fn rust_eh_personality() {
}

#[lang = "panic_impl"]
#[no_mangle]
pub extern fn rust_begin_panic(info: &PanicInfo) -> ! {
    unsafe { intrinsics::abort() }
}
```

In many cases, you may need to manually link to the `compiler_builtins` crate
when building a `no_std` binary. You may observe this via linker error messages
such as "```undefined reference to `__rust_probestack'```".

> ## More about the language items

Reviewer 2: Usually they're referred to without 'the'.

```suggestion
## More about language items
```

available in the executable to call. Normally these functions are provided by
the standard library, but without it you must define your own. These symbols
are called "language items", and they each have an internal name, and then a
signature that an implementation must conform to.

The first of these functions, `rust_eh_personality`, is used by the failure
mechanisms of the compiler. This is often mapped to GCC's personality function
(see the [libstd implementation][unwind] for more information), but crates
which do not trigger a panic can be assured that this function is never
called. The language item's name is `eh_personality`.

[unwind]: https://github.com/rust-lang/rust/blob/master/src/libpanic_unwind/gcc.rs

The second function, `rust_begin_panic`, is also used by the failure mechanisms of the
compiler. When a panic happens, this controls the message that's displayed on
the screen. While the language item's name is `panic_impl`, the symbol name is
`rust_begin_panic`.

Finally, a `eh_catch_typeinfo` static is needed for certain targets which
implement Rust panics on top of C++ exceptions.

## Well-known paths

> In several cases compiler finds specific item not by `lang` attribute. Instead

Reviewer 2: ```suggestion
In several cases, the compiler finds a specific item without using a `lang` attribute. Instead, the
```

(you may have to re-wrap the lines)

> item path is hardcored. For example compiler assumes `Iterator` trait to be

Reviewer 2: ```suggestion
item path is hardcoded. For example, the compiler assumes that the `Iterator` trait will be
```

> available as `core::iter::Iterator`. This only happens when item is required

Reviewer 2: ```suggestion
available as `core::iter::Iterator`. This only happens when an item is required
```

  - `bitxor_assign`: `library/core/src/ops/bit.rs`
  - `shl_assign`: `library/core/src/ops/bit.rs`
  - `shr_assign`: `library/core/src/ops/bit.rs`
  - `deref`: `library/core/src/ops/deref.rs`
  - `deref_mut`: `library/core/src/ops/deref.rs`
  - `index`: `library/core/src/ops/index.rs`
  - `index_mut`: `library/core/src/ops/index.rs`
  - `add`: `library/core/src/ops/arith.rs`
  - `sub`: `library/core/src/ops/arith.rs`
  - `mul`: `library/core/src/ops/arith.rs`
  - `div`: `library/core/src/ops/arith.rs`
  - `rem`: `library/core/src/ops/arith.rs`
  - `neg`: `library/core/src/ops/arith.rs`
  - `add_assign`: `library/core/src/ops/arith.rs`
  - `sub_assign`: `library/core/src/ops/arith.rs`
  - `mul_assign`: `library/core/src/ops/arith.rs`
  - `div_assign`: `library/core/src/ops/arith.rs`
  - `rem_assign`: `library/core/src/ops/arith.rs`
  - `eq`: `library/core/src/cmp.rs`
  - `ord`: `library/core/src/cmp.rs`
- Functions
  - `fn`: `library/core/src/ops/function.rs`
  - `fn_mut`: `library/core/src/ops/function.rs`
  - `fn_once`: `library/core/src/ops/function.rs`
  - `generator_state`: `library/core/src/ops/generator.rs`
  - `generator`: `library/core/src/ops/generator.rs`
- Opting out

>   - `unsafe_cell` (relaxes pointer provenance rules, allowing const-to-mut casts): `library/core/src/cell.rs`

Reviewer 2: Btw, you need to wrap this line.

# Lang items

The `rustc` compiler has certain pluggable operations, that is,
functionality that isn't hard-coded into the language, but is
implemented in libraries, with a special marker to tell the compiler
it exists. The marker is the attribute `#[lang = "..."]` and there are
various different values of `...`, i.e. various different 'lang
items'.

For example, `Box` pointers require two lang items, one for allocation
and one for deallocation. A freestanding program that uses the `Box`

> sugar for dynamic allocations via `malloc` and `free`:

Reviewer: Hmm, I'm not sure this is a great example because you can do this more easily with `#[global_allocator]`. Maybe use `#[lang = "start"]` instead? I think you can use that to have a graphical entrypoint instead of a console entrypoint on Windows. It would also simplify the example a lot I think - this shouldn't be a section on how to write an allocator ;)

    unsafe { intrinsics::abort() }
}
```

To override the compiler-inserted `main` shim, one has to disable it
with `#![no_main]` and then create the appropriate symbol with the
correct ABI and the correct name, which requires overriding the
compiler's name mangling too:

```rust,ignore
#![feature(lang_items, core_intrinsics)]
#![feature(start)]
#![no_std]
#![no_main]
use core::intrinsics;
use core::panic::PanicInfo;

// Pull in the system libc library for what crt0.o likely requires.
extern crate libc;

// Entry point for this program.
#[no_mangle] // ensure that this symbol is called `main` in the output
pub extern fn main(_argc: i32, _argv: *const *const u8) -> i32 {
    0
}

// These functions are used by the compiler, but not
// for a bare-bones hello world. These are normally
// provided by libstd.
#[lang = "eh_personality"]
#[no_mangle]
pub extern fn rust_eh_personality() {
}

#[lang = "panic_impl"]
#[no_mangle]
pub extern fn rust_begin_panic(info: &PanicInfo) -> ! {
    unsafe { intrinsics::abort() }
}
```

In many cases, you may need to manually link to the `compiler_builtins` crate
when building a `no_std` binary. You may observe this via linker error messages

> such as "```undefined reference to `__rust_probestack'```".

Reviewer: I don't think this section belongs in the dev-guide. I think there's already some info about this on https://os.phil-opp.com/freestanding-rust-binary/, and I think that's a better place - this doesn't tell you how the compiler works, just gives an example of a no_std executable, which seems like language documentation. If you want to add it to the reference, it might be a better fit there.

  - `div`: `libcore/ops/arith.rs`
  - `rem`: `libcore/ops/arith.rs`
  - `neg`: `libcore/ops/arith.rs`
  - `add_assign`: `libcore/ops/arith.rs`
  - `sub_assign`: `libcore/ops/arith.rs`
  - `mul_assign`: `libcore/ops/arith.rs`
  - `div_assign`: `libcore/ops/arith.rs`
  - `rem_assign`: `libcore/ops/arith.rs`
  - `eq`: `libcore/cmp.rs`
  - `ord`: `libcore/cmp.rs`
- Functions
  - `fn`: `libcore/ops/function.rs`
  - `fn_mut`: `libcore/ops/function.rs`
  - `fn_once`: `libcore/ops/function.rs`
  - `generator_state`: `libcore/ops/generator.rs`
  - `generator`: `libcore/ops/generator.rs`
- Opting out
  - `unsafe_cell` (relaxes pointer provenance rules, allowing const-to-mut casts): `libcore/cell.rs`
  - `manually_drop` (opts out of implicit destructor call): `libcore/mem/manually_drop.rs`
- Other
  - `coerce_unsized`: `libcore/ops/unsize.rs`
  - `drop`: `libcore/ops/drop.rs`
  - `drop_in_place`: `libcore/ptr.rs`
  - `clone`: `libcore/clone.rs`
  - `copy`: `libcore/marker.rs`
  - `send`: `libcore/marker.rs`
  - `sized`: `libcore/marker.rs`
  - `unsize`: `libcore/marker.rs`
  - `sync`: `libcore/marker.rs`
  - `phantom_data`: `libcore/marker.rs`
  - `discriminant_kind`: `libcore/marker.rs`

>   - `freeze`: `libcore/marker.rs`

Reviewer: This will *very* rapidly get out of date. Since you can easily rederive it with `rg 'lang = "'`, I don't think it needs to go in the guide.

Reviewer 2: (Also the filenames are wrong: it should be `library/core/src/...` not `libcore/...`)

Reviewer: Maybe it's better to link to https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/lang_items/enum.LangItem.html instead? That will always be up to date. It won't have the files in `library` but you can grep for those pretty easily.

Finally, a `eh_catch_typeinfo` static is needed for certain targets which
implement Rust panics on top of C++ exceptions.

## Well-known paths

In several cases compiler finds specific item not by `lang` attribute. Instead

> item path is hardcored. For example compiler assumes `Iterator` trait to be

Reviewer: ```suggestion
item path is hardcoded. For example compiler assumes `Iterator` trait to be
```
I didn't know this, thanks!

available as `core::iter::Iterator`. This only happens when item is required

> on early compilation stages (for example `Iterator` is used in for loops

Reviewer: ```suggestion
in early compilation stages (for example `Iterator` is used in `for` loops
```

desugaring).

> 

Reviewer 2: Actually, based on [looking at the output of `rustc +nightly -Z unpretty=hir`](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/why.20are.20.60Some.60.20and.20.60None.60.20language.20items.3F/near/219286210), `core::iter::Iterator` is *not* hardcoded when desugaring `for` loops. Do you know of a place where the language item's path *is* hardcoded, or should we just remove this section?

Reviewer 3: afaik we now try to use `QPath::LangItem` everywhere instead of using fully qualified paths.

https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/enum.QPath.html#variant.LangItem

I think this is probably out of date

Author: Yeah, you're right, the compiler no longer hardcodes paths. IIRC it was true some time ago, so the compiler was improved!

I'll verify if there are any well-known paths. If no I'll delete this section. Thanks!
